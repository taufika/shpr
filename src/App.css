.main-container {
  display: flex;
  height: calc(100vh - 80px);
}