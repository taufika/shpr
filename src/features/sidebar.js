import { createSlice } from '@reduxjs/toolkit'

export const sidebarSlice = createSlice({
  name: 'users',
  initialState: {
    isSidebarActive: false,
  },
  reducers: {
    toggleSidebar: (state) => {
      state.isSidebarActive = !state.isSidebarActive;
    },
  }
});

export const { toggleSidebar } = sidebarSlice.actions;

export default sidebarSlice.reducer;