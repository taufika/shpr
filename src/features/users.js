import { createSlice } from '@reduxjs/toolkit'

export const usersSlice = createSlice({
  name: 'users',
  initialState: {
    users: [],
    status: 'ready',
  },
  reducers: {
    setLoading: (state) => {
      state.status = 'loading';
    },
    setReady: (state) => {
      state.status = 'ready';
    },
    setError: (state) => {
      state.status = 'error';
    },
    assignUsers: (state, action) => {
      state.users = action.payload;
    },
  }
});

export const { assignUsers, setError, setReady, setLoading } = usersSlice.actions;

export default usersSlice.reducer;