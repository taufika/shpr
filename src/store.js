import { configureStore } from '@reduxjs/toolkit'
import usersReducer from './features/users';
import sidebarReducer from './features/sidebar';

export default configureStore({
  reducer: {
    users: usersReducer,
    sidebar: sidebarReducer,
  },
})
