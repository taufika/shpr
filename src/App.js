import './App.css';

import Navbar from './components/Navbar';

function App() {
  const user = {
    name: 'Shipper User',
  };

  return (
    <Navbar user={user}/>
  );
}

export default App;
