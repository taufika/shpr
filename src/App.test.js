import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import flushPromises from 'flush-promises';
import axios from 'axios';
import store from './store';
import App from './App';
import { assignUsers } from './features/users'

jest.mock('./components/UsersDisplay', () => (
  function UserDisplay() {
    // user display is rather complex so better test in the file
    return <div>User-Display-Mock</div>;
  }
));
jest.mock('axios', () => ({
  get: jest.fn(),
}));

const wrapped = () => <Provider store={store}><App /></Provider>;

const dummyResults = [
  {
    name: { first: 'Taufik', last: 'Akbar' },
  },
];

describe('Main Component: App', () => {
  afterEach(() => {
    axios.get.mockClear();
    store.dispatch(assignUsers([]));
    window.sessionStorage.removeItem('shprUsers');
  });

  it('Should render and match snapshot', () => {
    axios.get.mockResolvedValue({
      data: {
        results: dummyResults,
      },
    });
    const { container } = render(wrapped());
    expect(container).toMatchSnapshot();
  });

  it('Should call API to randomuser site', () => {
    axios.get.mockResolvedValue({
      data: {
        results: dummyResults,
      },
    });
    render(wrapped());
    expect(axios.get).toHaveBeenLastCalledWith('https://randomuser.me/api/?results=30');
  });

  it('Should use from session storage if exists', () => {
    axios.get.mockResolvedValue({
      data: {
        results: dummyResults,
      },
    });
    window.sessionStorage.setItem('shprUsers', JSON.stringify(dummyResults));
    render(wrapped());
    expect(axios.get).not.toHaveBeenCalledWith('https://randomuser.me/api/?results=30');
  });

  it('Should use from redux store if already exists', () => {
    axios.get.mockResolvedValue({
      data: {
        results: dummyResults,
      },
    });
    store.dispatch(assignUsers(dummyResults));
    render(wrapped());
    expect(axios.get).not.toHaveBeenCalledWith('https://randomuser.me/api/?results=30');
  });

  it('Should handle api failed', async () => {
    axios.get.mockRejectedValueOnce({});
    render(wrapped());
    await flushPromises();
    expect(store.getState().users.status).toBe('error');
  });

});
