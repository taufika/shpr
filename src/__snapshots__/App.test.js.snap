// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Component: App Should render and match snapshot 1`] = `
<div>
  <main
    class="app"
  >
    <nav
      class="navbar"
    >
      <button
        class="hamburger-menu"
        data-testid="navbar-hamburger"
      >
        <i
          class="fas fa-bars"
        />
      </button>
      <div
        class="logo"
      >
        <img
          alt="logo"
          class="logo"
          src="shipper-logo-black.png"
        />
      </div>
      <section
        class="user-info"
      >
        <h4
          class="hello"
        >
          Hello,
          <span
            class="user-name"
          >
            Shipper User
          </span>
        </h4>
        <div
          class="user-avatar"
        >
          <img
            alt="avatar"
            src="default_avatar.jpg"
          />
        </div>
      </section>
    </nav>
    <section
      class="main-container"
    >
      <div
        class="sidenav "
        data-testid="sidenav"
      >
        <div
          class="nav-item"
        >
          <i
            class="fas fa-home"
          />
          <span>
            Beranda
          </span>
        </div>
        <div
          class="nav-item active"
        >
          <i
            class="fas fa-user"
          />
          <span>
            Driver Management
          </span>
        </div>
        <div
          class="nav-item"
        >
          <i
            class="far fa-calendar"
          />
          <span>
            Pickup
          </span>
        </div>
      </div>
      <div>
        User-Display-Mock
      </div>
    </section>
  </main>
</div>
`;
